Tests/Unit/Hooks/HookProviderTest.php

```diff
@@ -0,0 +1,365 @@
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementorResetsCallCounterOnCreation()
+    {
+        TestingHookImplementor::$wasCalled = 1;
+
+        $this->createTestingHookImplementor();
+
+        self::assertSame(0, TestingHookImplementor::$wasCalled);
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementorImplementsHookHierachy()
+    {
+        $implementor = $this->createTestingHookImplementor();
+
+        self::assertInstanceOf(Hook::class, $implementor);
+        self::assertInstanceOf(TestingHookInterface::class, $implementor);
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementorTestHookMethodCanBeCalledAndReportsBeingCalled()
+    {
+        $implementor = $this->createTestingHookImplementor();
+
+        $implementor->testHookMethod();
+
+        self::assertSame(1, TestingHookImplementor::$wasCalled);
+    }
+
+    /*
+     * Tests concerning the TestingHookImplementor2
+     */
+
+    /**
+     * @test
+     */
+    public function hookImplementor2CanBeCreated()
+    {
+        self::assertInstanceOf(TestingHookImplementor2::class, $this->createTestingHookImplementor2());
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementor2ResetsCallCounterOnCreation()
+    {
+        TestingHookImplementor2::$wasCalled = 1;
+
+        $this->createTestingHookImplementor2();
+
+        self::assertSame(0, TestingHookImplementor2::$wasCalled);
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementor2ImplementsHookHierachy()
+    {
+        $implementor = $this->createTestingHookImplementor2();
+
+        self::assertInstanceOf(Hook::class, $implementor);
+        self::assertInstanceOf(TestingHookInterface::class, $implementor);
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementor2TestHookMethodCanBeCalledAndReportsBeingCalled()
+    {
+        $implementor = $this->createTestingHookImplementor2();
+
+        $implementor->testHookMethod();
+
+        self::assertSame(1, TestingHookImplementor2::$wasCalled);
+    }
+
+    /*
+     * Tests concerning the Hook object
+     */
+
+    /**
+     * @test
+     */
+    public function hookObjectForTestHookCanBeCreatedWithoutIndex()
+    {
+        self::assertInstanceOf(HookProvider::class, $this->createHookObject());
+    }
+
+    /**
+     * @test
+     */
+    public function hookObjectForNoInterfaceCannotBeCreated()
+    {
+        $this->expectException(\UnexpectedValueException::class);
+        $this->expectExceptionCode(1565089078);
+
+        $subject = new HookProvider('');
+    }
+
+    /**
+     * @test
+     */
+    public function hookObjectForNonexistantInterfaceCannotBeCreated()
+    {
+        $this->expectException(\UnexpectedValueException::class);
+        $this->expectExceptionCode(1565089078);
+
+        $subject = new HookProvider('\OliverKlee\Seminars\Tests\Unit\Fixtures\Hooks\TestingHookInterfaceDoesNotExist');
```

Comment:
Please generally use double backslashes in strinmg to avoid accidental escaping the next character.